crates/aranya-client/src/afc.rs

```diff
@@ -0,0 +1,437 @@
+    /// Convert AFC sequence object to `u64`.
+    pub fn to_u64(&self) -> u64 {
+        self.seq.to_u64()
+    }
+}
+
+/// Possible errors that could happen when using Aranya Fast Channels.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum AfcError {
+    /// Unable to seal datagram.
+    #[error("unable to seal datagram")]
+    Seal(aranya_fast_channels::Error),
+
+    /// Unable to open datagram.
+    #[error("unable to open datagram")]
+    Open(aranya_fast_channels::Error),
+
+    /// Unable to parse shared-memory path.
+    #[error("unable initialize shm")]
+    Shm(anyhow::Error),
+
+    /// No channel info found.
+    #[error("no channel info found")]
+    NoChannelInfoFound,
+
+    /// Error parsing control message.
+    #[error("failed to parse control message")]
+    InvalidCtrlMessage(postcard::Error),
+
+    /// An internal bug was discovered.
+    #[error(transparent)]
+    Bug(#[from] buggy::Bug),
+}
+
+/// AFC control message sent to a peer when creating a channel.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AfcCtrl {
+    data: Box<[u8]>,
+}
+
+impl AfcCtrl {
+    /// Convert AFC control message to bytes.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.data
+    }
+}
+
+/// Aranya Fast Channels handler for managing channels which allow encrypting/decrypting application data buffers.
+pub struct AfcChannels {
+    daemon: DaemonApiClient,
+    // TODO: don't use mutex for shm reader aranya-core#399
+    shm: Arc<Mutex<AfcShm>>,
+}
+
+// TODO: derive Debug on [`shm`] when [`AfcClient`] implements it.
+impl Debug for AfcChannels {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AfcChannels")
+            .field("daemon", &self.daemon)
+            .finish_non_exhaustive()
+    }
+}
+
+impl AfcChannels {
+    /// The number of additional octets required to encrypt
+    /// plaintext data.
+    pub const OVERHEAD: usize = AfcClient::<ReadState<CS>>::OVERHEAD;
+
+    pub(crate) fn new(daemon: DaemonApiClient, shm: Arc<Mutex<AfcShm>>) -> Self {
+        Self { daemon, shm }
+    }
+
+    /// Create a bidirectional AFC channel.
+    ///
+    /// Returns the channel object and a `ctrl` message to send to the peer.
+    pub async fn create_bidi_channel(
+        &self,
+        team_id: TeamId,
+        peer_id: DeviceId,
+        label_id: LabelId,
+    ) -> crate::Result<(AfcBidiChannel, AfcCtrl)> {
+        let (ctrl, channel_id) = self
+            .daemon
+            .create_afc_bidi_channel(context::current(), team_id, peer_id, label_id)
+            .await
+            .map_err(IpcError::new)?
+            .map_err(aranya_error)?;
+        Ok((
+            AfcBidiChannel {
+                daemon: self.daemon.clone(),
+                shm: self.shm.clone(),
+                channel_id,
+                label_id,
+            },
+            AfcCtrl { data: ctrl },
+        ))
+    }
+
+    /// Create a unidirectional AFC send-only channel.
```

Comment:
The docs for this + the recv variant are also too sparse. Eg, nothing says which peer is send only.


Additionally, it's not that the channel is send only or receive only. It's that one side can only send and the other side can only receive.


If you think about it for a minute that ends up being obvious, but I could see somebody who is unfamiliar with AFC getting confused.